### PR TITLE
Radar holes

### PR DESCRIPTION
**Pull request by cgsandford**
Plugin and CLI to fill in small holes within the radar composite, prior to extrapolation nowcasting.  (Met Office reviewer will find notebooks help with understanding / justifying some details.)

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- improver/cli/fill_radar_holes.py  1
- improver_tests/acceptance/test_fill_radar_holes.py  1
- improver_tests/nowcasting/utilities/test_FillRadarHoles.py  3
- improver/nowcasting/utilities.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292&bid=17735291)

File: improver/nowcasting/utilities.py [hunk 1]
```diff
@@ -36,17 +36,16 @@
 from cf_units import Unit
 
 from improver import BasePlugin
+from improver.utilities.spatial import distance_to_number_of_grid_cells
```
Comment (benfitzpatrick) [inline improver/nowcasting/utilities.py:39]: Codacy Issue found: [Unused distance_to_number_of_grid_cells imported from improver.utilities.spatial](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292)

File: improver/nowcasting/utilities.py [hunk 2]
```diff
@@ -103,6 +102,122 @@ def process(self, radar_data, coverage):
...
+                            x - r_speckle : x + r_speckle + 1,
+                        ]
+                    )
+                    if nbhood < p_masked:
+                        surroundings = log_rr[
+                            y - r_interp : y + r_interp + 1,
+                            x - r_interp : x + r_interp + 1,
+                        ]
+                        interpolated_points[y, x] = np.mean(
+                            surroundings[np.where(surroundings.mask == False)]
```
Comment (benfitzpatrick) [inline improver/nowcasting/utilities.py:172]: Codacy Issue found: [Comparison to False should be 'not expr'](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver/cli/fill_radar_holes.py  1
- improver_tests/acceptance/test_fill_radar_holes.py  1
- improver_tests/nowcasting/utilities/test_FillRadarHoles.py  3
- improver/nowcasting/utilities.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292&bid=17735291)

File: improver/nowcasting/utilities.py [hunk 3]
```diff
@@ -103,6 +101,122 @@ def process(self, radar_data, coverage):
         return cube_list.merge_cube()
 
 
+class FillRadarHoles(BasePlugin):
+    """Interpolate small "no data" regions in the radar composite domain
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply improver/nowcasting/utilities.py:105]: Agreed - I definitely need to be more consistent with this.  Will go over the docstrings as you suggest and rephrase.

File: improver/nowcasting/utilities.py [hunk 4]
```diff
@@ -103,6 +101,122 @@ def process(self, radar_data, coverage):
...
+            numpy.ma.MaskedArray:
+                Masked array of interpolated rainrates in log10(mm/h)
+        """
+        # minimum proportion of masked neighbours, below which a pixel is
+        # "speckle"
+        p_masked = 0.3
+        # radius of neighbourhood over which to search for masked neighbours
+        r_speckle = 4
+        # radius of neighbourhood from which to calculate interpolated values
+        r_interp = 2
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply improver/nowcasting/utilities.py:150]: No need for a check, but I can add further detail to the docstring about the justification of these values.  People should not be messing with these values without understanding what they mean, that would be very bad science.  Hence the hard-coding.

File: improver/nowcasting/utilities.py [hunk 5]
```diff
@@ -103,6 +101,122 @@ def process(self, radar_data, coverage):
...
+
+        interpolated_points = np.ma.MaskedArray(
+            np.full_like(log_rr, 0), mask=np.full_like(log_rr, True)
+        )
+
+        for y in range(r_speckle, log_rr.shape[0] - r_speckle - 1, 1):
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply improver/nowcasting/utilities.py:156]: Good point.  In the UK case we know we're working with a composite embedded in quite a large "no data" domain.  But I should document this more clearly for other users.

File: improver/nowcasting/utilities.py [hunk 6]
```diff
@@ -103,6 +101,122 @@ def process(self, radar_data, coverage):
...
+
+    @staticmethod
+    def _log_rr_to_rr(data):
+        """Convert a masked array of log rainrates into mm/h"""
+        result = np.where(np.isfinite(data), np.power(10, data), 0.0)
+        return np.ma.MaskedArray(result, mask=data.mask)
+
+    @staticmethod
+    def _find_and_interpolate_speckle(log_rr):
+        """Identify and interpolate "speckle" points.
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply improver/nowcasting/utilities.py:127]: OK, will do.

File: improver/nowcasting/utilities.py [hunk 7]
```diff
@@ -103,6 +101,122 @@ def process(self, radar_data, coverage):
...
+                            x - r_speckle : x + r_speckle + 1,
+                        ]
+                    )
+                    if nbhood < p_masked:
+                        surroundings = log_rr[
+                            y - r_interp : y + r_interp + 1,
+                            x - r_interp : x + r_interp + 1,
+                        ]
+                        interpolated_points[y, x] = np.mean(
+                            surroundings[np.where(~surroundings.mask)]
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply improver/nowcasting/utilities.py:171]: Thanks :+1:

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- improver/cli/fill_radar_holes.py  1
- improver_tests/acceptance/test_fill_radar_holes.py  1
- improver_tests/nowcasting/utilities/test_FillRadarHoles.py  3
- improver/nowcasting/utilities.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292&bid=17735291)

File: improver_tests/nowcasting/utilities/test_FillRadarHoles.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    assert np.allclose(result.data, OUTPUT_DATA[case])
+    assert np.array_equal(result.data.mask, OUTPUT_DATA[case].mask)
+
+
+def test_log_transform_reversability():
+    """Test that data in mm h-1 are reproduced when transforming to
+    and from log rainrates"""
+    log_rr = PLUGIN._rr_to_log_rr(RAIN_CUBE.data)
+    assert all(np.isnan(log_rr[np.where(RAIN_CUBE.data == 0)]))
+    rr = PLUGIN._log_rr_to_rr(log_rr)
```
Comment (benfitzpatrick) [inline improver_tests/nowcasting/utilities/test_FillRadarHoles.py:127]: Codacy Issue found: [Variable name "rr" doesn't conform to snake_case naming style](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292)

File: improver_tests/acceptance/test_fill_radar_holes.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+from . import acceptance as acc
+
+pytestmark = [pytest.mark.acc, acc.skip_if_kgo_missing]
+CLI = acc.cli_name_with_dashes(__file__)
+run_cli = acc.run_cli(CLI)
+
+
+def test_basic(tmp_path):
+    """Test interpolating radar holes"""
```
Comment (gavinevans) [inline improver_tests/acceptance/test_fill_radar_holes.py:45]: Maybe "Test filling radar holes by interpolation"?

File: improver/nowcasting/utilities.py [hunk 10]
```diff
@@ -103,6 +101,130 @@ def process(self, radar_data, coverage):
...
+        Returns:
+            iris.cube.Cube:
+                A masked cube with continuous coverage over the radar composite
+                domain, where missing data has been interpolated
+        """
+        # extract precipitation rate data in mm h-1
+        masked_radar_mmh = masked_radar.copy()
+        masked_radar_mmh.convert_units("mm h-1")
+
+        # interpolate "holes" in data
```
Comment (gavinevans) [inline improver/nowcasting/utilities.py:220]: Maybe 'fill "holes" in data'?

File: improver/nowcasting/utilities.py [hunk 11]
```diff
@@ -103,6 +101,130 @@ def process(self, radar_data, coverage):
...
+        """Convert a masked array of log rainrates into mm/h"""
+        result = np.where(np.isfinite(data), np.power(10, data), 0.0)
+        return np.ma.MaskedArray(result, mask=data.mask)
+
+    @staticmethod
+    def _find_and_interpolate_speckle(log_rr):
+        """Identify and interpolate "speckle" points, where "speckle" is defined
+        as areas of "no data" that are small enough to fill by interpolation
+        without affecting data integrity.  We would not wish to interpolate large
+        areas as this may gives false confidence in "no precipitation", where in
```
Comment (gavinevans) [inline improver/nowcasting/utilities.py:130]: Typo: `gives` => `give`

**Comment by cgsandford** [comment]
Wasn't sure which label to go with but @tjtg is going to rework the pytests unit tests using fixtures, so that we get a good "first example" of this unit test style for others to build on.  So this shouldn't be merged until I've pulled that in.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity increasing per file
==============================
- improver/cli/fill_radar_holes.py  1
- improver_tests/acceptance/test_fill_radar_holes.py  1
- improver_tests/nowcasting/utilities/test_FillRadarHoles.py  3
- improver/nowcasting/utilities.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292&bid=17735291)

File: improver_tests/nowcasting/utilities/test_FillRadarHoles.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
```
Comment (benfitzpatrick) [inline improver_tests/nowcasting/utilities/test_FillRadarHoles.py:1]: Codacy Issue found: [Module name "test_FillRadarHoles" doesn't conform to '(([a-z_][a-z0-9_]*)|([A-Z][a-zA-Z0-9]+))$' pattern](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292)

File: improver_tests/nowcasting/utilities/test_FillRadarHoles.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    """Test filling radar holes with a widespread mask,
+    interpolation should not be triggered in this case"""
+    plugin = FillRadarHoles()
+    rainrate.data[:11, :11] = np.nan
+    rainrate.data.mask = np.where(np.isfinite(rainrate.data.data), False, True)
+    result = plugin(rainrate)
+    check_fillradarholes(result, rainrate.data.copy())
+
+
+def test_log_transform_reversability(rainrate):
```
Comment (benfitzpatrick) [inline improver_tests/nowcasting/utilities/test_FillRadarHoles.py:113]: Codacy Issue found: [Function name "test_log_transform_reversability" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292)

File: improver_tests/nowcasting/utilities/test_FillRadarHoles.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    """Test that data in mm h-1 are reproduced when transforming to
+    and from log rainrates"""
+    plugin = FillRadarHoles()
+    log_rr = plugin._rr_to_log_rr(rainrate.data)
+    np.testing.assert_array_equal(log_rr[np.where(rainrate.data == 0)], np.nan)
+    rr = plugin._log_rr_to_rr(log_rr)
```
Comment (benfitzpatrick) [inline improver_tests/nowcasting/utilities/test_FillRadarHoles.py:119]: Codacy Issue found: [Variable name "rr" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292)

File: improver/nowcasting/utilities.py [hunk 15]
```diff
@@ -103,6 +101,130 @@ def process(self, radar_data, coverage):
...
+        # radius of neighbourhood over which to search for masked neighbours
+        r_speckle = 4
+        # radius of neighbourhood from which to calculate interpolated values
+        r_interp = 2
+
+        interpolated_points = np.ma.MaskedArray(
+            np.full_like(log_rr, 0), mask=np.full_like(log_rr, True)
+        )
+
+        for y in range(r_speckle, log_rr.shape[0] - r_speckle - 1, 1):
```
Comment (benfitzpatrick) [inline improver/nowcasting/utilities.py:163]: Codacy Issue found: [Variable name "y" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292)

File: improver/nowcasting/utilities.py [hunk 16]
```diff
@@ -103,6 +101,130 @@ def process(self, radar_data, coverage):
...
+            for x in range(r_speckle, log_rr.shape[1] - r_speckle - 1, 1):
```
Comment (benfitzpatrick) [inline improver/nowcasting/utilities.py:164]: Codacy Issue found: [Variable name "x" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292)

File: improver_tests/acceptance/test_fill_radar_holes.py [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+Tests for the fill-radar-holes CLI
+"""
+
+import pytest
+
+from . import acceptance as acc
+
+pytestmark = [pytest.mark.acc, acc.skip_if_kgo_missing]
+CLI = acc.cli_name_with_dashes(__file__)
+run_cli = acc.run_cli(CLI)
```
Comment (benfitzpatrick) [inline improver_tests/acceptance/test_fill_radar_holes.py:41]: Codacy Issue found: [Constant name "run_cli" doesn't conform to '(([A-Z_][A-Z0-9_]*)|(__.*__))$' pattern](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292)

File: improver_tests/acceptance/test_fill_radar_holes.py [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+# POSSIBILITY OF SUCH DAMAGE.
+"""
+Tests for the fill-radar-holes CLI
+"""
+
+import pytest
+
+from . import acceptance as acc
+
+pytestmark = [pytest.mark.acc, acc.skip_if_kgo_missing]
```
Comment (benfitzpatrick) [inline improver_tests/acceptance/test_fill_radar_holes.py:39]: Codacy Issue found: [Constant name "pytestmark" doesn't conform to '(([A-Z_][A-Z0-9_]*)|(__.*__))$' pattern](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver/cli/fill_radar_holes.py  1
- improver_tests/acceptance/test_fill_radar_holes.py  1
- improver_tests/nowcasting/utilities/test_FillRadarHoles.py  3
- improver/nowcasting/utilities.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292&bid=17735291)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver/cli/fill_radar_holes.py  1
- improver_tests/acceptance/test_fill_radar_holes.py  1
- improver_tests/nowcasting/utilities/test_FillRadarHoles.py  3
- improver/nowcasting/utilities.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292&bid=17735291)

File: improver_tests/acceptance/test_fill_radar_holes.py (hunk 9 above)
Comment (cgsandford) [reply improver_tests/acceptance/test_fill_radar_holes.py]: Oops - missed this one, sorry.  Fixed now.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver/cli/fill_radar_holes.py  1
- improver_tests/acceptance/test_fill_radar_holes.py  1
- improver_tests/nowcasting/utilities/test_FillRadarHoles.py  3
- improver/nowcasting/utilities.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292&bid=17735291)

File: improver_tests/nowcasting/utilities/test_FillRadarHoles.py [hunk 19]
```diff
@@ -41,8 +40,9 @@
 from ...set_up_test_cubes import set_up_variable_cube
 
 
-def create_masked_rainrate_data():
-    """Create a masked data array of rain rates in mm/h"""
+@pytest.fixture(name="rainrate")
+def rainrate_fixture() -> Cube:
```
Comment (gavinevans) [inline improver_tests/nowcasting/utilities/test_FillRadarHoles.py:44]: @tjtg I notice that [type hints](https://docs.python.org/3/library/typing.html) have been used with these pytest fixtures Is this a general style to be following for pytest fixtures? I'm not sure if we've previously used type hints in IMPROVER.

Comment (tjtg) [reply improver_tests/nowcasting/utilities/test_FillRadarHoles.py:44]: I used the type hints here as a way to include the type information, without expanding to a longer multiline docstring containing a `Returns` section. A `Returns` section would be close to a duplicate of the existing docstring summary, so it seemed unnecessary.

Pytest doesn't have any requirement for type hints, it was just a convenient way to express the information about the fixture. Having type hints may be helpful for other fixture functions, but I didn't intend using it here as a hard requirement for others to follow.

Comment (gavinevans) [reply improver_tests/nowcasting/utilities/test_FillRadarHoles.py:44]: OK, thanks for the explanation.

**Comment by cgsandford** [comment]
Replaced by #1250

**Pull request closed without merging**